feature: docker registry configuration

ricardofbarros opened the pull request:
Feature:
- external docker registry configuration 
- docker registry authentication based on user-password

Closes: [polyaxon/#291](https://github.com/polyaxon/polyaxon/issues/291) and [polyaxon/#257](https://github.com/polyaxon/polyaxon/issues/257)

I probably need some help with the integration tests, to do something similar to postgres, there is an external IP to the postgres but not sure if the postgres instance is up or not, or if it's just a mocked IP.

cc:/ @mouradmourafiq 


PS: Sorry for opening the issue on the wrong repo, my bad.

mouradmourafiq commented on polyaxon/values.yaml:
```diff
@@ -424,7 +424,7 @@ postgresql:
   #     postgresUser:
   #     postgresPassword:
   #     postgresDatabase:
-  enabled: true
```
Can you update the the comment in [line 421](https://github.com/polyaxon/polyaxon-chart/pull/23/files#diff-43bc6296fbc9a23e12a4cce5c0f46e0eL421)

ricardofbarros replied:
Done 👍

mouradmourafiq commented on polyaxon/values.yaml:
```diff
@@ -490,6 +490,10 @@ rabbitmq:
             topologyKey: "kubernetes.io/hostname"
 
 docker-registry:
```
It would be nice to have a similar comment as in [line 419](https://github.com/polyaxon/polyaxon-chart/pull/23/files#diff-43bc6296fbc9a23e12a4cce5c0f46e0eL419)

ricardofbarros replied:
Done 👍

mouradmourafiq commented:
Thank you for this great contribution, I am in the process of cutting a new release, if you think you can make the changes, I will wait before releasing to include this as well, let me know.

mouradmourafiq commented on polyaxon/templates/cm.yaml:
```diff
@@ -192,7 +192,13 @@ data:
+  {{- end }}
+  {{- if (index .Values "docker-registry").registryUser }}
+  POLYAXON_REGISTRY_USER: {{ (index .Values "docker-registry").registryUser }}
+  POLYAXON_REGISTRY_PASSWORD: {{ (index .Values "docker-registry").registryPassword }}
```
This should probably be a secret instead of env var?

ricardofbarros replied:
@mouradmourafiq I agree with you, I have created a secret for docker-registry when `install: false`, but the reason I used that env var is because I think `polyaxon` only looks to that env variable. I couldn't find a away to use the secret like we have for postgreSQL with `POLYAXON_K8S_DB_SECRET_NAME`. So maybe if we go through this secret feature, I should also do a PR for polyaxon/polyaxon to use a k8s secret for the docker registry when available?

ricardofbarros replied:
@mouradmourafiq here^

mouradmourafiq replied:
@ricardofbarros sorry for the late answer, the secret should mimic as you said the behavior of postgreSQL. We actually do not pass the name of the registry secret to the platform, but we can change that, by adding a new env var `POLYAXON_K8S_REGISTRY_SECRET_NAME` similar to this [one](https://github.com/polyaxon/polyaxon-chart/blob/master/polyaxon/templates/cm.yaml#L38).

It should point out to the default registry secret, you will find it in the docker-registry [subchart](https://github.com/polyaxon/polyaxon-chart/blob/master/polyaxon/charts/docker-registry/templates/secrets.yaml), otherwise to the one you just created when `install is false`.

For passing `POLYAXON_REGISTRY_PASSWORD` you need to create a new entry similar to the postgres, redis, and rabbitmq, [here](https://github.com/polyaxon/polyaxon-chart/blob/master/polyaxon/templates/partials/_envs.tpl)

ricardofbarros replied:
@mouradmourafiq I fixed your suggestions in the last commit, I'll mark this as solved. Will continue discussion on main thread.

mouradmourafiq commented on polyaxon/values.yaml:
```diff
@@ -490,6 +490,10 @@ rabbitmq:
+  install: true
+  registryUser:
+  registryPassword:
+  externalRegistryHost:
```
I would just call it `registryHost:`.

ricardofbarros replied:
I went with `externalRegistryHost` to follow the same convention as the external postgres

mouradmourafiq replied:
Sorry my bad, I did not pay attention 👍

ricardofbarros commented:
@mouradmourafiq I went with `install` by following the the gitlab helm chart, they do something similar to this but happy to change to `deploy`.

Thanks a lot for the brief explanation how the integration tests works, will fix those and the comments you posted.

I think I can have this done by today if you are okay to wait that long

mouradmourafiq commented:
@ricardofbarros thanks for the quick reply, let's keep it `install`, I don't mind  :)

I went ahead and released the v0.3.5, I think it's better to include this contribution in the v0.3.6, the v0.3.5 already includes too many changes already.

If you need any help with tests let me know.

ricardofbarros commented:
@mouradmourafiq I fixed the things you pointed out. The only remaining doubts are the `secret` if changes are needed or not on polyaxon repository. And the integration tests.

I tested locally on my machine and I'm not able to replicate the issue on Travis CI.

I ran `python3 ci/validate/helm.py --values=values1.yml --output-dir=tmp_template1` and the lint went OK in my machine.

I tried to google the error but to no avail: `[ERROR] templates/: rendering template failed: reflect: Zero(nil)`

mouradmourafiq commented:
@ricardofbarros I will try your branch tomorrow to see if I can spot the issue, and I will let you know.

ricardofbarros commented:
@mouradmourafiq I was able to fix the bug now I got other bug to fix but I'll be fine :) 

I just need some help regarding the secret, could you answer me on the discussion you opened?

mouradmourafiq commented:
@ricardofbarros good to hear that you managed to find what was the issue, I just made a comment on the registry secret name and password in the thread.

ricardofbarros commented:
@mouradmourafiq I think I have fixed everything. 

Although the integration tests are still failing, but I think the reason they are failing is because I rebased this feature branch with master. 

The integration tests are failing(hanging) on the same script and the same command:
```
./ci/test-deploy.sh
...
...
...
+helm install --name polyaxon-test --namespace polyaxon ./polyaxon/ -f ./ci/test-config.yml
Error: timed out waiting for the condition
```

Should I reset my branch to the latest git tag?

ricardofbarros commented on polyaxon/templates/secrets.yaml:
```diff
@@ -77,3 +77,20 @@ data:
   {{ end }}
 {{- end -}}
+---
+{{ if (index .Values "docker-registry").auth.password }}
```
@mouradmourafiq I disagree with this. The way I see this is that auth is separate feature (that can be enabled). If you check the `values.yaml` I separated the `user` and `password` from `external` key and created a `auth` key.

So that's why I don't check `install` like you do on postgresql.

But let me know if you don't agree with this.

dyngts commented:
Hi @ricardofbarros, thanks for enabling this feature.

Is there any docs to getting started?

ricardofbarros commented:
Hey @dyngts, there isn't any docs for this feature but look into `values.yaml` on `docker-registry` key and you will find a comment how to use this feature.

But I think this feature wasn't cut into a feature so you'll have to wait until it's released.

Also @mouradmourafiq should this feature be documented on the website? Care to give some pointers what I need to edit to add this information?

mouradmourafiq commented:
@dyngts as @ricardofbarros  said this feature is not released yet. A release is coming this weekend (v0.3.6), and the feature will be in beta and you can try it and report bugs/enhancement.

@ricardofbarros we should create a new section similar to [postgres High availability](https://docs.polyaxon.com/customization/postgresql_ha/). So I would say, we can start with an `.md` file under the [docs](https://github.com/polyaxon/polyaxon/tree/master/docs/templates/customization) folder where we can document how the users can use this feature. If you would like to add this file it would be awesome, you can start with the simplest documentation possible (may be just moving the values to the markdown?). After merging the `.md` file, I will add it to the doc's index/config just before cutting a release.
